src/runtime_src/doc/toc/xrt_native_apis.rst

```diff
@@ -0,0 +1,618 @@
+ 
+   
+**C++**: The equivalent C++ code
+
+In C++ the ``xrt::kernel`` class provides **overloaded operator ()** to execute the kernel with a comma-separated list of arguments.  
+
+.. code:: c++
+      :number-lines: 50
+   
+           // 1st kernel execution
+           auto run = kernel(buf_a, buf_b, scalar_1); 
+           run.wait();
+    
+           // 2nd kernel execution with just changing 3rd argument
+           run.set_arg(2,scalar_2); // Arguments are specified starting from 0 
+           run.start();
+           run.wait();
+
+The above c++ code block is demonstrating 
+  
+  - The kernel execution using the ``xrt::kernel()`` operator with the list of arguments that returns a xrt::run object. This is an asynchronous API and returns after submitting the task.    
+  - The method ``xrt::run::wait`` is used to block the current thread until the current execution is finished. 
+  - The method ``xrt::run::set_arg`` is used to set one or more kernel argument(s) before the next execution. In the example above, only the last (3rd) argument is changed.  
+  - The method ``xrt::run::start`` is used to start the next kernel execution with new argument(s).   
+
+Other kernel execution related APIs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Run handle/object supports few other use-cases. 
+
+**Obtaining the run handle/object before execution**: In the above example we have seen a run handle/object is obtained when the kernel is executed (kernel execution returns a run handle/object). However, a run handle can be obtained even before the kernel execution. The flow is as below
+
+    - Open a Run handle (or object) by API ``xrtRunOpen`` (in C++ ``xrt::run`` constructor with a kernel argument). There is no kernel execution associated with this run handle/object yet
+    - Set the kernel arguments associated for the next execution by ``xrtRunSetArg`` (in C++ method ``xrt::run::set_arg``). 
+    - Execute the kernel by ``xrtRunStart`` (in C++ method ``xrt::run::start``).
+    - Wait for the execution finish by ``xrtRunWait`` (C++: ``xrt::run::wait``). 
+
+**Timeout while wait for kernel finish**: The API ``xrtRunWait`` blocks the current thread until the kernel execution finishes. However, a timeout supported API ``xrtRunWaitFor`` is also provided . The timeout number can be specified using a millisecond unit.
+
+In C++, the timeout facility can be used by the same ``xrt::run::wait(unsigned int timeout_ms=0)`` method by providing a millisecond number as an argument. 
+
+**Asynchronous update of the kernel arguments**: The API ``xrtRunSetArg`` (C++: ``xrt::run::set_arg``) is synchronous to the kernel execution. This API can only be used when kernel is in the IDLE state and before the start of the next execution. An asynchronous version of this API (only for edge platform) ``xrtRunUpdateArg`` (in C++ method ``xrt::run::update_arg``) is provided to change the kernel arguments asynchronous to the kernel execution. 
```

Comment:
"to change the kernel arguments asynchronous to the kernel execution." what is the consistency memory model for this?

Reply:
Just to give you some hints: you have a 256 bit kernel register updated while the kernel is running and using it and the interface cannot update it atomically. How the implementation or the user is handling the tearing?

Reply:
xrtRunUpdateArg is messed up completely.   We are fully aware of the missing atomicity (sic?) so behavior is as defined in kernel, or APIs should be used for word sized scalars only.   Someone insisted this API was needed, I still have not seen the use case.

Reply:
OK, so at least we totally agree :-)
If you liked metastability, you will love our new `xrtRunUpdateArg` ;-)

Reply:
nothing to do for this PR, resolving